bigbluebutton/greenlight
# Add support for handling of expired rooms

## Description


Rake tasks were added that can be used to handle room expiration. In total three tasks were added:

- The task _remove_expired_rooms_  can be used to mark rooms as deleted in the database that are either not used for a specified time or that were never used and where created longer ago than the mentioned specified period of time.
- The task _permanently_remove_deleted_rooms_ can be used to permanently remove rooms that are longer than a specified period of time marked as deleted.
- The task _notify_of_expiring_rooms_ can be used to notify users via E-Mail of rooms that are potentially expiring within a specified period of time if they are not used again. The notification can happen in English or German.

To execute these tasks automatically periodically (i.e., once a day) they are scheduled by Cron jobs which are running in a separate Docker container and which are defined using the _Whenever_  gem. These jobs are parameterized using environment variables that can be defined within the _.env_-file and that are explained within the _sample.env_-file. If these environment variables are not defined, no Cron jobs are executed. If they are defined, they can be used to implement the following room expiration handling:

- Users are notified via E-Mail of rooms that would expire within a long-term period of time (e.g., ten weeks) if they are not used again until the denoted date.
- Users are notified again via E-Mail of rooms that were not used since the long-term expiration notification and that would thous expire within a short-term period of time (e.g., three days) if they are not used again until the denoted date.
- Rooms that were not used for a specified period of time (e.g., 13 months) or that were never used and where created longer ago than the mentioned specified period of time are marked as deleted in the database.
- Rooms that were longer than a specified period of time (e.g., 12 months) in the deleted state are permanently removed from the database.

## Testing Steps


The defined rake tasks are tested within [room_spec.rb](spec/tasks/room_spec.rb). The scheduling can be tested manually by starting the Docker Compose setup and checking the database for expected modifications and checking the added mailhog client (separate container added for testing purposes in the Docker Compose setup) for expected notifications.

## Screenshots (if appropriate):

Expiration notification:

![expiration_notification](https://user-images.githubusercontent.com/110479684/196398301-458da41f-094e-4133-9ca6-2d5cb19425ed.png)
